music21/opensheetmusicdisplay/osmd.py

```diff
@@ -0,0 +1,189 @@
+                var s = document.createElement( 'script' );
+                // alternative to local file is using a CDN
+                // s.setAttribute( 'src', "https://cdn.jsdelivr.net/npm/opensheetmusicdisplay@0.3.1/build/opensheetmusicdisplay.min.js" );
+                //s.setAttribute( 'src', "{{script_dir}}" );
+                function oncompleted(){
+                    window.define = _define
+                    console.log("loaded OSMD for the first time",opensheetmusicdisplay)
+                    resolve(opensheetmusicdisplay);
+                };
+                {{script_command}}
+                
+            }) 
+        }
+        loadOSMD().then((OSMD)=>{
+            console.log("loaded OSMD",OSMD)
+            var div_id = "{{DIV_ID}}";
+                console.log(div_id)
+            document.querySelector('#'+div_id).innerHTML = "";
+            window.openSheetMusicDisplay = new OSMD.OpenSheetMusicDisplay(div_id);
+            openSheetMusicDisplay
+                .load({{data}})
+                .then(
+                  function() {
+                    console.log("rendering data")
+                    openSheetMusicDisplay.render();
+                  }
+                );
+        })
+        """ \
+        .replace('{{DIV_ID}}',div_id) \
+        .replace('{{data}}',json.dumps(xml)) \
+        .replace('{{script_dir}}',pathname2url(script_dir))
+
+        if offline is True:
+            script_content = open(script_dir).read()
+            script = script.replace('{{script_command}}',"""
+                s.type = 'text/javascript';
+                s.text="""+json.dumps(script_content)+""";
+                document.body.appendChild( s ); // browser will try to load the new script tag
+                oncompleted();
+                """)
+        else:
+            script = script.replace('{{script_command}}',"""
+                s.setAttribute( 'src', 'https://cdn.jsdelivr.net/npm/opensheetmusicdisplay@0.3.1/build/opensheetmusicdisplay.min.js' );
+                s.onload=oncompleted;
+                document.body.appendChild( s ); // browser will try to load the new script tag
+                """)
+
+        return script
+
+    @staticmethod
+    def addDefaultPartName(score):
+        # If no partName is present in the first instrument, OSMD will display the ugly 'partId'
+        allInstruments = list(score.getInstruments(returnDefault=False, recurse=True))
+        
+        if len(allInstruments)==0:
+            # print("adding default inst")
+            defaultInstrument = Piano()
+            defaultInstrument.instrumentName = 'Default'
+            score.insert(None, defaultInstrument)
+            score.coreElementsChanged()
+        elif not allInstruments[0].instrumentName: 
+            # print("adding instrumentName")
+            # instrumentName must not be '' or None
+            allInstruments[0].instrumentName = 'Default'
+            score.coreElementsChanged()
```

Comment:
The elements haven't changed in this case, only the name of the instrument, so no need to call coreElementsChanged() -- coreElementsChanged() only needs to be run if another `core*` function has been called.  All general music21 stream commands will call it if necessary and not call it if not.